pyro/infer/inspect.py

```diff
@@ -0,0 +1,191 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Callable, Dict, List, Optional
+
+import torch
+
+import pyro
+import pyro.poutine as poutine
+from pyro.poutine.messenger import Messenger
+from pyro.poutine.util import site_is_subsample
+
+
+def is_sample_site(msg):
+    if msg["type"] != "sample":
+        return False
+    if site_is_subsample(msg):
+        return False
+
+    # Ignore masked observations.
+    if msg["is_observed"] and msg["mask"] is False:
+        return False
+
+    # Exclude deterministic sites.
+    fn = msg["fn"]
+    while hasattr(fn, "base_dist"):
+        fn = fn.base_dist
+    if type(fn).__name__ == "Delta":
+        return False
+
+    return True
+
+
+class RequiresGradMessenger(Messenger):
+    def __init__(self, predicate=lambda msg: True):
+        self.predicate = predicate
+        super().__init__()
+
+    def _pyro_post_sample(self, msg):
+        if is_sample_site(msg):
+            if self.predicate(msg):
+                msg["value"].requires_grad_()
+            elif not msg["is_observed"] and msg["value"].requires_grad:
+                msg["value"] = msg["value"].detach()
+
+
+def get_dependencies(
+    model: Callable,
+    model_args: Optional[tuple] = None,
+    model_kwargs: Optional[dict] = None,
+) -> Dict[str, List[str]]:
+    r"""
+    Infers metadata about a conditioned model. Metadata includes
+
+    -   `prior_dependencies` is a dict mapping downstream latent and observed
+        sites to the dictionaries mapping upstream latent sites on which the
+        depend to sets of plates over which dependencies are independent.
+        Dependencies follow the original model order.
```

Comment:
It's probably worth adding an even simpler example of this data structure inline in these two bullet points and spelling out what it means, if only to save ourselves a bit of hassle if we want to refactor this in 6 months.

e.g.: "...original model order. For example, if `prior_dependencies == {"a": {"a": set()}, "b": {"b": {"i"}, "a": set()}}`, then `a` is a sample site with no ancestors and no plates, while each entry `b_i` of sample site `b` is conditionally independent along plate `i` given a value of `a`."

and correspondingly below, "Posterior dependencies are reversed from the model order. For example, if we condition on sample site `b` in the example above, then `posterior_dependencies == {"a": {"a": set(), "b": set()}}`: `b` no longer appears as a key because it is observed, while global variable `a` now depends on all entries of `b` in plate `i`."